### PR TITLE
chore: land the #2247 round-9 review findings on v2/main

### LINKED ISSUE #2264: The testing → test-servers hand-off lands on the wrong half of the procedure
Follow-up to #2247 / #2261. Copilot filed a ninth review round **after that PR merged**, so none of it is on `v2/main`. Five findings, three of them substantive.

### The headline one: the hand-off succeeds by the metric and still delivers the wrong instructions

#2247 got `testing → test-servers` to 100%, but `test-servers`'s body documents **one** way to use a fixture: run `server-composable.js --config <name>.json` as a second process and connect the Inspector to it. That is the by-hand path.

An integration test does not work that way. It builds the server **in-process** and owns its lifecycle:

```ts
// clients/web/src/test/integration/mcp/inspectorClient-excluded-tools.test.ts
import { createTestServerHttp, createTestServerInfo, createEchoTool }
  from "@modelcontextprotocol/inspector-test-server";

const started = createTestServerHttp({ serverInfo: …, tools: [createEchoTool()] });
await started.start();
// … afterEach → started.stop()
```

No process is spawned, no JSON config is read, no showcase file is picked. So the model loads the skill — the thing #2247 spent eight review rounds securing — and gets the API/lifecycle guidance it needs **nowhere**, plus a nudge toward an external config and process it should not use.

**This is worth stating plainly: a green hand-off number is not evidence the hand-off is useful.** The eval can only observe that skill B was loaded. Whether B then answers the question is outside what any `chain` case measures, and #2247 did not notice because 100% looked like success.

### The rest

| # | Finding | File |
| --- | --- | --- |
| 2 | Heading still asks *"does the test need a real server?"*, which the body's own hand-rolled-server exception contradicts. Should ask about a `test-servers/` fixture. | `.claude/skills/testing/SKILL.md` |
| 3 | *"imports"* is too broad a dependency test — `src/test/core/mcp/test-server-scope.test.ts` imports `@modelcontextprotocol/inspector-test-server` to unit-test its scope middleware as a pure function and never starts a server. The condition is **starts, spawns, or configures** a fixture. | same |
| 4 | The worked example attributes the 33% → 100% rise to imperative wording *and* placement independently; the edit changed both at once, so only the combination is established. | `docs/skill-authoring.md` |
| 5 | `CHAIN_THRESHOLD` comment generalises "a well-shaped pointer clears 0.8" from one two-case run of a multi-variable edit. | `scripts/skill-eval.mjs` |

### State

**All five are already written**, as an uncommitted patch against `0d947ad5` — 4 files, +81/-20, including a new **"Two ways to use a fixture — pick the right one first"** section in `test-servers` with the in-process recipe and a pointer to the reference test. It is saved locally and needs re-applying on a fresh branch off `v2/main`.

⚠️ **It is not verified.** The one `skills:eval` run against that patch reported **31/63 first-move and 0/2 hand-off** — a collapse across skills the patch does not touch, which is the signature of a rate limit rather than a regression (`docs/skill-authoring.md` warns the eval "goes red on a rate limit"). **That has to be re-run before anything is concluded**; if it reproduces, the patch is at fault and finding 1's new section is the first suspect, since it is the only large body addition.

### Acceptance

- The five findings fixed on `v2/main`.
- `RUNS=5 npm run skills:eval` re-run clean: hand-off stays above 50%, no first-move regression. The 0/2 reading explained — rate limit or real.
- A reader who lands in `test-servers` from an integration-test prompt finds the in-process API, not only the two-process one.

### PR DESCRIPTION
**Pull request by cliffhall**
Closes #2264

Copilot filed a ninth review round on #2247 **after that PR merged**, so none of it reached `v2/main`. All five findings are addressed here. Docs, skill bodies and one comment only — no runtime code changes.

### 1 — The hand-off landed on the wrong half of the procedure

`testing` → `test-servers` scored 100%, and still delivered the wrong instructions. `test-servers` documented exactly one way to use a fixture: run `server-composable.js --config <name>.json` as a second process and point the Inspector at it. An integration test does not work that way — it builds the server **in-process** from `@modelcontextprotocol/inspector-test-server` and owns its lifecycle.

- `test-servers` gains a **"Two ways to use a fixture — pick the right one first"** section: the in-process recipe (`createTestServerHttp` / `start()` / `stop()` in an `afterEach`), a pointer to the reference test (`inspectorClient-excluded-tools.test.ts`), and the four mechanics that bite — one barrel for the factories, stop in `afterEach` or the port leaks, read `started.url` rather than assuming a port, and era is a constructor option (`modern: {}`) not a config file.
- ⚠️ kept: the barrel aliases to `test-servers/build/index.js`, so the stale-build hazard applies to the in-process path in full.
- `## Run one` → **`## Run one by hand (two processes)`**, so the manual path names itself instead of reading as *the* path.
- `testing`’s connecting bullet now says plainly that an automated test builds in-process and does **not** spawn, and points at the right half.

### 2 — The heading contradicted the body

`## Before you write it: does the test need a real server?` → `does the test use a `test-servers/` fixture?`. The body’s own hand-rolled-server exception (`inspectorClient-malformed-list.test.ts`, `listSalvage-era.test.ts`) is a real server with no `test-servers/` dependency, so the old heading asked the wrong question.

### 3 — "imports" was too broad a dependency test

`src/test/core/mcp/test-server-scope.test.ts` imports the package to unit-test `createScopeCheckMiddleware` as a pure function and never stands a server up. The condition is now **starts, spawns, or configures**, with that test named as the counterexample.

### 4 — The worked example over-attributed

`docs/skill-authoring.md` presented imperative wording and placement as two independently-established levers. The #2247 edit changed both in one revision and only the combination was measured; the split is now labelled the plausible mechanism rather than a result, with the two runs that would separate them named as unpaid-for.

### 5 — `CHAIN_THRESHOLD` generalised from one run

The comment read "a well-shaped pointer clears 0.8" off a single two-case run of that same multi-variable edit. It now reads as an existence proof that the ceiling is above 0.8 — not a property of pointers in general, and not grounds for raising the floor.

### The lesson, recorded where it can be found again

**A green hand-off number proves skill B loaded, never that B answered the question.** A `chain` case can observe one thing. Whether the target then says something useful to the caller who arrived through the pointer is outside what it measures — which is exactly how #2247 shipped a 100% hand-off into the wrong half of a procedure. That is now stated in `AGENTS.md`’s chain-case rules and in the `docs/skill-authoring.md` checklist, with the instruction it implies: when a chain goes green, read the target’s body **as the caller who arrives through A**.

### Verification

- `npm run local:gate` — green (exit 0) in a clean worktree with a real `npm install`.
- `npm run verify:skills` — OK: 10 skills, 9 model-invoked, listing 3234/4000 chars (unchanged; no description was touched).
- `RUNS=5 npm run skills:eval` — running; the result, and the explanation of the issue’s unexplained 0/2 reading, are posted as a comment below before merge.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018zwYWanPhvgu6wiKoeGg4N

### CODE REVIEW

File: .claude/skills/test-servers/SKILL.md [hunk 1]
```diff
@@ -17,6 +17,92 @@ and what the broken build did — is
 [`docs/test-servers.md`](../../../docs/test-servers.md). This skill is how to
 run one.
 
+## Two ways to use a fixture — pick the right one first
+
+A fixture is used in **one of two shapes**, and almost everything below is about
+the second. Establish which one you are in before reading further, because the
+config file, the protocol-era table and the staleness hazard belong to only one
+of them.
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/test-servers/SKILL.md]: Fixed in a83712d5 — you are right, and the wrong half was the *build*, not the config.

All three shapes resolve `test-servers/build/`: the two automated ones through the `@modelcontextprotocol/inspector-test-server` alias in `vitest.shared.mts`, and the manual one by running the emitted `.js` directly. So the intro now scopes only the showcase **config file and the protocol-era table** to the manual path, and carries an explicit ⚠️ saying the build applies to every shape and that `Build first` is not manual-path guidance.

`Build first` itself opened with "The servers are spawned as real subprocesses," which gave the wrong reason for the import path — that is your third comment, and it is fixed in the same commit.

File: .claude/skills/test-servers/SKILL.md [hunk 2]
```diff
@@ -17,6 +17,92 @@ and what the broken build did — is
...
+
+⚠️ **The barrel is an alias to the BUILD, not to the source** —
+`vitest.shared.mts` maps `@modelcontextprotocol/inspector-test-server` to
+`test-servers/build/index.js`. So the `Build first` section applies to this path
+in full, including the stale-build hazard: an edit to `test-servers/src` that is
+not rebuilt is invisible to an in-process test exactly as it is to a spawned one.
+
 ## Build first
 
 The servers are spawned as real subprocesses, so the build output must exist:
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/test-servers/SKILL.md]: Fixed in a83712d5. `Build first` now opens:

> Every shape above resolves generated output — an automated test imports the barrel, which is **aliased to `test-servers/build/index.js`**, and the stdio and manual paths run emitted `.js` as real subprocesses. So the build must exist whichever one you are in:

That is the internally-consistent version: the import path is named as a reason of its own rather than being covered by a subprocess claim that does not hold for it.

File: .claude/skills/test-servers/SKILL.md [hunk 3]
```diff
@@ -17,6 +17,92 @@ and what the broken build did — is
...
+A fixture is used in **one of two shapes**, and almost everything below is about
+the second. Establish which one you are in before reading further, because the
+config file, the protocol-era table and the staleness hazard belong to only one
+of them.
+
+- **In-process — an automated test.** The test *constructs* the server from the
+  `@modelcontextprotocol/inspector-test-server` API and owns its lifecycle. No
+  subprocess is spawned, no JSON config is read, and no showcase config is
+  picked. **This is what an integration test does**, and it is the shape you
+  want whenever the caller is a test rather than a person.
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/test-servers/SKILL.md]: Fixed in a83712d5 — confirmed against the repo before changing it. `getTestMcpServerCommand()` has 20 call sites across five `clients/web/src/test/integration/mcp/**` files and the whole `clients/cli/__tests__` suite, so this was not an edge case; the section was telling people not to use the established pattern.

"Two ways" is now **"Three ways to use a fixture"**, opening with a table (shape / who / where the server runs / config file) and a subsection per automated shape. The stdio one states the three things that differ:

- **The transport owns the child, not you** — there is no `start()`/`stop()` pair, so the `afterEach` that matters is `client.disconnect()`.
- **No config is selected and none can be.** `getTestMcpServerCommand()` returns `node <test-servers/build/test-server-stdio.js>`, which runs that server's default config — so the showcase table and era guidance genuinely do not apply here either, but for a different reason than in-process.
- **It is still the build**, with the same staleness hazard.

Reference tests named for it: `inspectorClient-response-rejected.test.ts` and `clients/cli/__tests__/methods.test.ts`, plus a note that `e2e.test.ts` spawns the built CLI which in turn spawns the fixture.

The "every fixture-backed integration test in this repo follows" claim on the in-process reference test is narrowed to "when the server is built in-process," with a pointer to the stdio subsection.

File: .claude/skills/testing/SKILL.md [hunk 4]
```diff
@@ -23,8 +23,16 @@ ways to depend on one, and they need different halves of that skill:
...
+  against a server. These need the whole procedure — the staleness hazard, and
+  then whichever half matches how the server is stood up.
+  ⚠️ **An automated test builds the server IN-PROCESS; it does not spawn one.**
+  It calls `createTestServerHttp(...)` / `.start()` / `.stop()` from
+  `@modelcontextprotocol/inspector-test-server` and owns the lifecycle — no
+  subprocess, no JSON config, no showcase config to pick and no protocol-era
+  table to consult (era is a constructor option). That API and its reference
+  test are the **"Two ways to use a fixture"** section of `test-servers`; the
+  showcase config and era guidance there are the *manual* path and do not apply
+  to you. Read the right half.
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/testing/SKILL.md]: Fixed in a83712d5. The "does not spawn" instruction is gone; what replaces it is the two automated shapes and when to pick each:

- **In-process HTTP** — `createTestServerHttp(...)` / `.start()` / `.stop()`, when the case needs HTTP or SSE, a specific tool set, or the modern handler.
- **Spawned stdio** — `getTestMcpServerCommand()` into a stdio transport or the built CLI, which spawns it; when stdio is the point and the default tool set suffices.

The property they actually share — and the one this bullet needed in the first place — is that **neither picks a showcase config**, so the config/era half of `test-servers` applies to neither while its build warning applies to both. That is what it now says, ending in "Read the right third" instead of "the right half."

**Comment by cliffhall** [comment]
## Review round 1 — all four addressed (a83712d5)

Replied inline on each thread; mirrored here because those replies go hidden once the fix is pushed. No "Suppressed comments" block in the review body.

| # | Finding | Done |
| --- | --- | --- |
| 1 | Staleness hazard scoped to the manual path, contradicting the new lines | ✅ |
| 2 | Automated stdio tests were misclassified as in-process-only | ✅ |
| 3 | `Build first` opens with a subprocess reason that excludes the import path | ✅ |
| 4 | `testing`’s "does not spawn" instruction is too broad | ✅ |

**#2 was the substantive one and it was right.** I checked before changing anything: `getTestMcpServerCommand()` has 20 call sites across five `clients/web/src/test/integration/mcp/**` files and the whole `clients/cli/__tests__` suite. The section was telling callers not to use the repo’s established stdio pattern.

So **"Two ways" is now "Three ways to use a fixture"**, with a shape/who/where-it-runs/config-file table and a subsection for the spawned-stdio path — the transport owns the child (`client.disconnect()`, not `stop()`), no config can be selected because that entry point runs the stdio default, it is still the build, and two reference tests are named. `testing` now presents the two automated shapes and when to pick each, and the property they share is the one that bullet actually needed: **neither picks a showcase config**.

#1 and #3 are the same mistake from two sides, and the wrong half was the *build*, not the config. All three shapes resolve `test-servers/build/` — two through the `vitest.shared.mts` alias, one by running the emitted `.js` — so only the showcase config and era table are scoped to the manual path now, and `Build first` names the import path as a reason of its own.

### Verification on the final tree

- **`RUNS=5 npm run skills:eval` — clean.** `63/63` first-move cases at or above 80%, `2/2` hand-off cases above 50% (100% and 80%). **That also settles #2264’s unexplained `31/63` + `0/2` reading: it was a rate limit, not a regression.** Two full-suite runs on this branch reproduced neither number.
- **`npm run local:gate` — green on the first commit; the re-run after round 1 went red on one 5s timeout in `SkillsScreen.test.tsx`, which is a load flake, not this change.** This PR touches **zero** files under `clients/` (`git diff origin/v2/main...HEAD --name-only` is five markdown/tooling files), that file took 91s for 111 tests in the red run, and re-running it alone passes 111/111. Every other stage was green.
- `npm run verify:skills` — OK; listing 3234/4000 chars, unchanged. No description was edited, so no first-move trigger surface moved.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018zwYWanPhvgu6wiKoeGg4N

File: .claude/skills/test-servers/SKILL.md [hunk 5]
```diff
@@ -17,9 +17,149 @@ and what the broken build did — is
...
+
+A fixture is used in **one of three shapes**, and most of what follows is about
+the third. Establish which one you are in before reading further, because the
+showcase **config file and the protocol-era table** belong to that one alone.
+
+| Shape | Who | Server runs | Config file |
+| --- | --- | --- | --- |
+| **In-process HTTP** | an automated test | inside the test process, built from the API | none — options are constructor args |
+| **Spawned stdio** | an automated test | a child process the transport starts | none — the stdio fixture runs its default config |
+| **Two processes** | a person, by hand | a terminal you started | yes — a showcase `--config <name>.json` |
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/test-servers/SKILL.md]: Fixed in fb5aeddd — you are right, and the fault was the axis, not the row. Verified both call sites: `scripts/smoke-web-elicitation.mjs` spawns `server-composable.js --config` directly, and `scripts/lib/mcp-app-flow.mjs` (`startMcpAppServer`) does it for `smoke:web:app`, `smoke:web:tabs` and `pack:verify`.

Round 1 split the shapes by "automated vs by hand," which put a config-driven smoke on the wrong side of the line. The section now says plainly that **the cut is how the server is stood up, not who is driving**, with an explicit ⚠️ saying so and why. The third row is **Spawned composable HTTP**, and its config/era guidance belongs to the *shape*, whoever starts it:

| Shape | Server runs | Config file | Consumers |
| --- | --- | --- | --- |
| In-process HTTP | inside the test process | none | integration tests |
| Spawned stdio | child the transport starts | none — default config | integration tests, the CLI suites |
| **Spawned composable HTTP** | child started with `--config` | **yes** | **the web smokes, `pack:verify`, and a person by hand** |

The `Who` column that caused this is gone, replaced by `Consumers`. The prose bullet names both call sites and says the whole skill applies to them; `Run one by hand` is now described as that shape driven by a person rather than as a different path.

File: .claude/skills/testing/SKILL.md [hunk 6]
```diff
@@ -23,8 +23,26 @@ ways to depend on one, and they need different halves of that skill:
...
   end-to-end test that connects; a smoke that drives a connected flow; a
   coverage gap only reachable over a real connection; reproducing a reported bug
-  against a server. These need the whole procedure — which showcase config,
-  which protocol era, and the staleness hazard.
+  against a server. These need the whole procedure — the staleness hazard, and
+  then whichever half matches how the server is stood up.
+  ⚠️ **An automated test does not pick a showcase config, whichever way it
+  stands the server up.** Two shapes, and both come from the
+  `@modelcontextprotocol/inspector-test-server` API rather than from
+  `test-servers/configs/`:
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/testing/SKILL.md]: Fixed in fb5aeddd. The blanket "an automated test does not pick a showcase config" is gone. `testing` now routes by what is driving, with three destinations rather than a binary:

- **An integration or CLI test → in-process HTTP.** No showcase config, no era table.
- **An integration or CLI test where stdio is the point → spawned stdio.** Default config, nothing to pick *and nothing to override* — so a case needing a specific tool set is an in-process HTTP test instead.
- **A smoke → spawned composable HTTP, and it IS config-driven.** `smoke:web:elicitation`, `smoke:web:app`, `smoke:web:tabs` and `pack:verify` all spawn `server-composable.js --config <name>.json`, so the showcase-config and protocol-era guidance applies to them in full — with the point stated explicitly, since it is the one this review caught twice: **being automated does not exempt a smoke from it.**

What applies to all three is the build warning, and that is the one line kept in common.

**Comment by cliffhall** [comment]
## Review round 2 — both addressed, plus the suppressed one (fb5aeddd)

Replied inline on both threads; mirrored here, and the suppressed finding is answered here because it has no thread.

**Round 1 fixed the missing shape but kept the wrong axis, and that is what this round caught.** Splitting by "automated vs by hand" put the config-driven smokes on the wrong side of the line. Verified both call sites before changing anything: `scripts/smoke-web-elicitation.mjs:143-147` spawns `server-composable.js --config` directly, and `scripts/lib/mcp-app-flow.mjs:92-102` (`startMcpAppServer`) does it for `smoke:web:app`, `smoke:web:tabs` and `pack:verify`. Those are automated *and* need the full config/era guidance.

| # | Finding | Done |
| --- | --- | --- |
| 1 | Config-driven path is not manual-only; smokes spawn it too | ✅ |
| 2 | `testing`’s "an automated test does not pick a showcase config" is false for smokes | ✅ |
| S | *(suppressed)* stale "an automated test builds the server in-process instead" under `Run one by hand` | ✅ |

**The fix is the axis.** `test-servers` now states outright that the cut is **how the server is stood up, not who is driving**, and carries a ⚠️ saying why that axis was wrong. The third shape is **Spawned composable HTTP**, whose `Consumers` are "the web smokes, `pack:verify`, **and** a person by hand" — the `Who` column that caused this is gone. `Run one by hand` is now that same shape driven by a person, which also disposes of the suppressed comment: the stale sentence it quoted is deleted.

`testing` routes to three destinations instead of a binary — integration/CLI test → in-process HTTP; stdio-is-the-point → spawned stdio (default config, nothing to pick *or override*, so a specific tool set means in-process instead); **smoke → composable config, and the era guidance applies in full**, with "being automated does not exempt a smoke from it" said in as many words.

### Verification

- **`npm run local:gate` — green (exit 0), zero `FAIL` lines.** This is the clean re-run; the earlier red was the `SkillsScreen.test.tsx` load flake explained in the round-1 summary, and it did not recur.
- `npm run verify:skills` — OK; listing 3234/4000 chars, unchanged.
- `RUNS=5 npm run skills:eval` was clean on the round-1 tree (63/63 first-move, 2/2 hand-off). This round edits the same two bodies; **I will re-run it on this tree and post the numbers before merge** rather than carry the earlier run forward, since the whole point of #2264 is that a hand-off number describes the body it was measured against.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018zwYWanPhvgu6wiKoeGg4N

File: .claude/skills/test-servers/SKILL.md [hunk 7]
```diff
@@ -17,9 +17,160 @@ and what the broken build did — is
...
+the showcase **config file and the protocol-era table** belong to that one alone.
+
+⚠️ **The cut is how the server is stood up, not who is driving.** Automated and
+by-hand is the wrong axis: the composable-config shape has *both* kinds of
+consumer, and a smoke that spawns it needs every bit of the config and era
+guidance a person at two terminals does.
+
+| Shape | Server runs | Config file | Consumers |
+| --- | --- | --- | --- |
+| **In-process HTTP** | inside the test process, built from the API | none — options are constructor args | integration tests |
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/test-servers/SKILL.md]: Fixed in 98750905 — verified both, and both are right. `scripts/smoke-cli.mjs:310-322` imports `createTestServerHttp` and starts it **in the smoke process** (deliberately: it reads back `getRecordedRequests()` to prove the CLI actually put the header on the wire), and `clients/cli/__tests__/cli.test.ts:377` does the same behind a forwarding proxy.

Every row’s `Consumers` is now filled in rather than left to the reader:

| Shape | Consumers |
| --- | --- |
| In-process HTTP | integration tests, CLI tests, **`smoke:cli`** |
| Spawned stdio | integration tests, the CLI suites, **`smoke:cli`**, **`smoke:tui`** |
| Spawned composable HTTP | the **web** smokes, `pack:verify`, and a person by hand |

— narrowed to "the **web** smokes" on the third row, with a ⚠️ under the table saying the thing that keeps going wrong in as many words: **"a smoke" is not a shape.** `smoke:cli` uses the first two and `smoke:tui` the second; pick by the row, never by the caller’s category.

Two supporting edits fell out of it: the in-process bullet no longer says "the test" constructs it (a smoke does too, and that is the example given), and the stdio bullet now covers *naming* the built entry in a `--catalog` without connecting, which is what `smoke:cli`’s connect checks and `smoke:tui` actually do.

File: .claude/skills/testing/SKILL.md [hunk 8]
```diff
@@ -23,8 +23,29 @@ ways to depend on one, and they need different halves of that skill:
...
+    `getTestMcpServerCommand()` handed to a stdio transport or to the built CLI,
+    which spawns it. A subprocess *is* started, but it runs the stdio fixture's
+    **default** config, so there is still nothing to pick — and nothing to
+    override, so if the case needs a specific tool set it is an in-process HTTP
+    test instead.
+  - **A smoke → spawned composable HTTP, and it IS config-driven.**
+    `smoke:web:elicitation`, `smoke:web:app`, `smoke:web:tabs` and `pack:verify`
+    all spawn `server-composable.js --config <name>.json`. **The showcase-config
+    and protocol-era guidance applies to you in full** — being automated does
+    not exempt a smoke from it.
```
[the remark this answers is not in the document]
Comment (cliffhall) [reply .claude/skills/testing/SKILL.md]: Fixed in 98750905. Round 2 traded one wrong axis for a narrower one — routing on the word "smoke" — and `smoke:cli` is the counterexample: an in-process `createTestServerHttp` for the header round-trip (`smoke-cli.mjs:310-322`), and the built stdio entry written into a `--catalog` for the connect checks (`:119`, `:231-232`). Neither picks a config.

The heading is narrowed to **"A config-driven web smoke, or `pack:verify` → spawned composable HTTP"**, and it is followed by:

> ⚠️ **"A smoke" is not a shape, so do not route by that word.** `smoke:cli` uses the first two — an in-process `createTestServerHttp` for the header round-trip, and the built stdio entry in a `--catalog` for the connect checks — and `smoke:tui` uses the stdio entry alone. Only the web smokes above are config-driven. Pick by what the caller actually stands up.

So both of `smoke:cli`’s shapes are named at the routing point, not just implied by the two bullets above it.

**Comment by cliffhall** [comment]
## Review round 3 — both addressed, plus the suppressed one (98750905)

Replied inline on both threads; mirrored here, and the suppressed finding is answered here because it has no thread.

**Round 2 traded one wrong axis for a narrower one: routing on the word "smoke".** `smoke:cli` is the counterexample, and it uses *two* shapes, neither config-driven — `smoke-cli.mjs:310-322` starts an in-process `createTestServerHttp` (deliberately: it reads back `getRecordedRequests()` to prove the CLI put the header on the wire), and `:119`/`:231-232` write the built stdio entry into a `--catalog`. `clients/cli/__tests__/cli.test.ts:377` starts an in-process one too.

| # | Finding | Done |
| --- | --- | --- |
| 1 | Consumer list omits `smoke:cli` / CLI tests from the in-process row | ✅ |
| 2 | `testing` routes every smoke to the config-driven path | ✅ |
| S | *(suppressed)* "starts, spawns, or configures" excludes the build-only `smoke:tui` documented right above it | ✅ |

Every row’s `Consumers` is now filled in — in-process HTTP gains CLI tests and `smoke:cli`, spawned stdio gains `smoke:cli` and `smoke:tui`, and the composable row is narrowed to the **web** smokes — with the recurring mistake stated outright under the table and again at `testing`’s routing point: **"a smoke" is not a shape; pick by what the caller actually stands up.** The stdio bullet also now covers *naming* the built entry in a catalog without connecting, which is what those two smokes do.

**The suppressed one was a real self-contradiction** and worth calling out: `testing` documented `smoke:tui` as a build-only dependency in one bullet and then gave a criterion — "starts, spawns, or configures" — that excluded it. It now reads **"starts, spawns, configures, or hands a built entry to the subject under test"**, with the clause tied back to that bullet.

### Verification

- **`npm run local:gate` — green (exit 0), zero `FAIL` lines.**
- ⚠️ One note for the record: an earlier run of the gate went red on **five 5000ms timeouts** across `SkillsScreen.test.tsx` and `skillsVerification.test.ts` (both #2248). Not this change — this PR touches **zero** files under `clients/` (`git diff origin/v2/main...HEAD --name-only` is five markdown/tooling files), the failures were timeouts rather than assertions, and re-running both files alone passes **156/156**. That cluster looks load-sensitive at the default 5s `testTimeout` independently of this branch; I am filing it separately rather than smuggling a fix in here.
- `npm run verify:skills` — OK; listing 3234/4000 chars, unchanged.
- `RUNS=5 npm run skills:eval` — re-running on this tree now; numbers posted before merge.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018zwYWanPhvgu6wiKoeGg4N

**Comment by cliffhall** [comment]
## Rounds 4 and 5 — two consecutive zero-comment approvals ✅

No new comments in either round, so the review has converged. Round 5:

> 🟢 **Approval recommended** — All five requested documentation corrections are addressed consistently, with prior review findings incorporated and no unresolved defects found.

Four rounds of findings, every one of them real and every one from the same family: **the taxonomy kept being cut on the wrong axis.** Worth recording, because it is the same failure mode #2264 is about.

| Round | What was wrong | The axis it was cut on |
| --- | --- | --- |
| 1 | Missing the spawned-stdio shape entirely | in-process vs. manual |
| 2 | Config-driven **smokes** on the manual side of the line | automated vs. by hand |
| 3 | `smoke:cli` routed to the config path; it uses the other two | the word "smoke" |
| 4, 5 | — | how the server is stood up |

Only the last one survives, and it is the one the skill now states outright.

## Final verification, on the merge tree

### `RUNS=5 npm run skills:eval`

```
Hand-off (14 turns)
PASS 100%  testing → test-servers     Write an integration test that exercises tool listing end to end.
PASS 100%  testing → test-servers     Add end-to-end coverage for the tool-list pagination path.

2/2 hand-off cases above 50%.
```

**100% / 100%, up from 100% / 80% on the round-1 tree** — the pointer got stronger as the target's body got more accurate, which is the direction you would want but is not something this PR set out to claim.

**First-move: 62/63 at or above 80%, with one caveat I am not going to paper over.** `pr-flow`'s *"I've finished the fix for issue 2071. Take it through to a pull request."* scored **60%** on that run. It is noise, and here is the evidence rather than the assertion:

- It scored **100% on two earlier full runs of this same branch**.
- A focused `RUNS=5 npm run skills:eval -- pr-flow` on the final tree returns **7/7 at 100%**, that case included.
- This PR edits **no description**. The listing is byte-identical at 3234/4000 chars, so the first-move trigger surface did not move at all — and `pr-flow`'s body is not touched either.

That is exactly the swing `docs/skill-authoring.md` warns about for a single `RUNS=5` sample, and this PR adds the honest reading of it rather than treating the green number as the verdict.

**It also settles #2264's open question.** The issue reported an unexplained `31/63` first-move and `0/2` hand-off against the original patch and flagged the new `test-servers` section as the first suspect if it reproduced. **It did not reproduce** — three full `RUNS=5` runs on this branch, across three different versions of that section, and none came close to those numbers. It was a rate limit.

### Gate

`npm run local:gate` — **green (exit 0), zero `FAIL` lines** on the final tree.

The intermittent `SkillsScreen` / `skillsVerification` timeouts seen on two earlier runs are filed as **#2327** (Todo / Medium) with the five-run record and the isolation evidence. Not fixed here, deliberately: this PR touches nothing under `clients/`, and smuggling an unrelated test fix into it would make both changes harder to read.

`npm run verify:skills` — OK; 10 skills, 9 model-invoked, listing 3234/4000.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_018zwYWanPhvgu6wiKoeGg4N
